Add support for i586-unknown-netbsd as target.

he32 opened the pull request:
This restricts instructions to those offered by Pentium, to support e.g. AMD Geode.

There is already an entry for this target in the NetBSD platform support page at

  src/doc/rustc/src/platform-support/netbsd.md

...so this should forestall its removal.

Additional fixes are needed for some vendored modules, this is the changes in the rust compiler core itself.

he32 committed:
Add support for i586-unknown-netbsd as target.

This restricts instructions to those offered by Pentium,
to support e.g. AMD Geode.

There is already an entry for this target in the NetBSD
platform support page at

  src/doc/rustc/src/platform-support/netbsd.md

...so this should forestall its removal.

Additional fixes are needed for some vendored modules, this
is the changes in the rust compiler core itself.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6642670768/job/18047964664) [(plain)](https://github.com/rust-lang/rust/commit/6642b4b1e2f2edec71cdf3fabef8fcdc8b8517a7/checks/18047964664/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
Removing intermediate container f63e14f1d448
 ---> ba4143d1fc43
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> df6c9ecccff7
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-kd2yhiis/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 20.3 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 47.3 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.4 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container 65ae85b0d788
 ---> 3ef6ef444e27
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 4dabf9f338ec
 ---> 4dabf9f338ec
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 568b8006367c
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 957501fcaac0
 ---> c95f57563e60
Successfully built c95f57563e60
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:c95f57563e607fcb346c1024a7d093d39b96c8b4832a883d25f31c25bfc77e0c
Uploading finished image sha256:c95f57563e607fcb346c1024a7d093d39b96c8b4832a883d25f31c25bfc77e0c to https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
c95f57563e60   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
4dabf9f338ec   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
3ef6ef444e27   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
df6c9ecccff7   10 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
ba4143d1fc43   12 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      2 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      2 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      2 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f3a9c042f10>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
sha256:c95f57563e607fcb346c1024a7d093d39b96c8b4832a883d25f31c25bfc77e0c
---
DirectMap4k:      204736 kB
DirectMap2M:     8183808 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/eb03d40a9c946bbf57c655ed7d26dc44416ca4ad/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-eb03d40a9c946bbf57c655ed7d26dc44416ca4ad-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 25.83s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_target/src/spec/i586_unknown_netbsd.rs at line 1:
-use crate::spec::{Cc, Lld, LinkerFlavor, StackProbeType, Target, TargetOptions};
+use crate::spec::{Cc, LinkerFlavor, Lld, StackProbeType, Target, TargetOptions};
 pub fn target() -> Target {
 pub fn target() -> Target {
     let mut base = super::netbsd_base::opts();
##[error]Diff in /checkout/compiler/rustc_target/src/spec/i586_unknown_netbsd.rs at line 5:
     base.cpu = "pentium".into();
     base.max_atomic_width = Some(64);
-    base.pre_link_args
-        .entry(LinkerFlavor::Gnu(Cc::Yes, Lld::No))
-        .or_default()
-        .push("-m32".into());
+    base.pre_link_args.entry(LinkerFlavor::Gnu(Cc::Yes, Lld::No)).or_default().push("-m32".into());
     base.stack_probes = StackProbeType::Call;
     Target {
     Target {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/i586_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/aarch64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/powerpc64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/s390x_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/arm_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/hexagon_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/x86_64_sun_solaris.rs" "/checkout/compiler/rustc_target/src/spec/powerpc64le_unknown_linux_musl.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Oct 25 15:33:33 UTC 2023
  network time: Wed, 25 Oct 2023 15:33:33 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

oli-obk commented:
cc @jethrogb

bjorn3 commented on compiler/rustc_llvm/build.rs:
```diff
@@ -258,6 +258,12 @@ fn main() {
     {
         println!("cargo:rustc-link-lib=z");
     } else if target.contains("netbsd") {
+        // Building for i586 or i686, we need -latomic for 64-bit atomics
```
Does that use lock-free atomics? Also is this needed for the rust standard library as well?

he32 replied:
To be perfectly honest, I have not looked under the hood for that.  The `-latomic` library comes from `gcc`.  I believe this is the only available solution for this platform since LLVM stubbornly insists on relying on 64-bit atomic operations.  i586, which is "Pentium"-class CPUs, do, as far as I'm aware, not have any native-to-the-CPU 64-bit atomic instructions, so I cling on to what appears to be the only reasonable alternative.

As for whether this is also needed for the rust standard library?  I don't *think* so -- I think these were the only i586-specific patches that I could find in our `pkgsrc` package (though I still need to submit pull requests for some of the vendored components, such as the rust `libc`).  The resulting rust compiler has been used to cross-build and build rust for this target for quite a while, and the resulting compiler has successfully been used to build e.g. firefox (natively).

bjorn3 replied:
If the cpu doesn't support 64bit atomics, `max_atomic_width` should probably be 32bit. We guarantee that atomics are lock-free and declare them as unavailable when they wouldn't be. We don't emulate them using mutexes.

he32 replied:
That's a fair point as far as rust itself goes, and I see that the target spec says 64, so this sends me back to the test bench to see if the build will succeed with that set to 32 instead.

However, that said, the above modification concerns LLVM's reliance on 64-bit atomics, and that also needs to be dealt with one way or the other.

he32 replied:
That said, the `i586_pc_nto_qnx700.rs` target spec both says `llvm_target: "i586-pc-unknown".into(),` and sets atomic width to 64 bits.  How it manages to do that I do not know...

he32 replied:
I changed `max_atomic_width` to 32 and re-tried the build of `rust` from scratch.

The initial attempt fell on its face with

```
     Running `/usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/bootstrap/debug/rustc --crate-name crossbeam_utils --edition=2018 /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/crossbeam-utils/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --cfg 'feature="std"' -Zunstable-options --check-cfg 'values(feature, "default", "loom", "nightly", "std")' --check-cfg 'names()' --check-cfg 'values()' -C metadata=e0b5eeecf1f9537e -C extra-filename=-e0b5eeecf1f9537e --out-dir /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/i586-unknown-netbsd/stage1-rustc/i586-unknown-netbsd/release/deps --target i586-unknown-netbsd -L dependency=/usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/i586-unknown-netbsd/stage1-rustc/i586-unknown-netbsd/release/deps -L dependency=/usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/i586-unknown-netbsd/stage1-rustc/release/deps --extern cfg_if=/usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/i586-unknown-netbsd/stage1-rustc/i586-unknown-netbsd/release/deps/libcfg_if-6e2d2163a7e2c47b.rmeta --cap-lints allow --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(parallel_compiler)' '--check-cfg=values(no_btreemap_remove_entry)' '--check-cfg=values(crossbeam_loom)' '--check-cfg=values(span_locations)' '--check-cfg=values(rustix_use_libc)' '--check-cfg=values(emulate_second_only_system)' '--check-cfg=values(windows_raw_dylib)' -Zmacro-backtrace -Clink-args=-Wl,-z,origin -Clink-args=-Wl,-rpath,/usr/pkg/lib -Zunstable-options -Csplit-debuginfo=off -Zunstable-options '-Wrustc::internal' -Cprefer-dynamic -Z binary-dep-depinfo`
error[E0432]: unresolved imports `core::sync::atomic::AtomicI64`, `core::sync::atomic::AtomicU64`
  --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/crossbeam-utils/src/lib.rs:79:49
   |
79 |             pub(crate) use core::sync::atomic::{AtomicI64, AtomicU64};
   |                                                 ^^^^^^^^^  ^^^^^^^^^ no `AtomicU64` in `sync::atomic`
   |                                                 |
   |                                                 no `AtomicI64` in `sync::atomic`
   |
help: a similar name exists in the module
   |
79 |             pub(crate) use core::sync::atomic::{AtomicI32, AtomicU64};
   |                                                 ~~~~~~~~~
help: a similar name exists in the module
   |
79 |             pub(crate) use core::sync::atomic::{AtomicI64, AtomicU32};
   |                                                            ~~~~~~~~~

error[E0412]: cannot find type `AtomicU64` in module `core::sync::atomic`
    --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/crossbeam-utils/src/atomic/consume.rs:78:14
     |
78   | impl_atomic!(AtomicU64, u64);
     |              ^^^^^^^^^ help: a struct with a similar name exists: `AtomicU16`
     |
    ::: /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/library/core/src/sync/atomic.rs:2031:9
     |
2031 |         pub struct $atomic_type {
     |         ----------------------- similarly named struct `AtomicU16` defined here

error[E0412]: cannot find type `AtomicI64` in module `core::sync::atomic`
    --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/crossbeam-utils/src/atomic/consume.rs:80:14
     |
80   | impl_atomic!(AtomicI64, i64);
     |              ^^^^^^^^^ help: a struct with a similar name exists: `AtomicI16`
     |
    ::: /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/library/core/src/sync/atomic.rs:2031:9
     |
2031 |         pub struct $atomic_type {
     |         ----------------------- similarly named struct `AtomicI16` defined here

Some errors have detailed explanations: E0412, E0432.
For more information about an error, try `rustc --explain E0412`.

Did not run successfully: exit status: 1
```

However, with this change to the vendored `crossbeam-utils` crate(s), I get further along:

```
--- vendor/crossbeam-utils/no_atomic.rs.orig    2023-10-03 04:12:09.000000000 +0000
+++ vendor/crossbeam-utils/no_atomic.rs
@@ -28,15 +28,16 @@ const NO_ATOMIC_64: &[&str] = &[
     "armv6k-nintendo-3ds",
     "avr-unknown-gnu-atmega328",
     "hexagon-unknown-linux-musl",
+    "i586-unknown-netbsd",
     "m68k-unknown-linux-gnu",
     "mips-unknown-linux-gnu",
     "mips-unknown-linux-musl",
```

(Though, mysteriosly, my first attempt ended up with a SEGV, but re-starting the build is still progressing beyond that point.) ... But ... the build still fails with

```
error[E0412]: cannot find type `AtomicU64` in module `atomic`
    --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:282:30
     |
282  | pub struct AtomicU64(atomic::AtomicU64);
     |                              ^^^^^^^^^ help: a struct with a similar name exists: `AtomicU16`
     |
    ::: /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/library/core/src/sync/atomic.rs:2031:9
     |
2031 |         pub struct $atomic_type {
     |         ----------------------- similarly named struct `AtomicU16` defined here
     |
note: struct `crate::write::zip_writer::AtomicU64` exists but is inaccessible
    --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:282:1
     |
282  | pub struct AtomicU64(atomic::AtomicU64);
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible

error[E0433]: failed to resolve: could not find `AtomicU64` in `atomic`
   --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:286:22
    |
286 |         Self(atomic::AtomicU64::new(v))
    |                      ^^^^^^^^^
    |                      |
    |                      could not find `AtomicU64` in `atomic`
    |                      help: a struct with a similar name exists: `AtomicU16`
    |
note: struct `crate::write::zip_writer::AtomicU64` exists but is inaccessible
   --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:282:1
    |
282 | pub struct AtomicU64(atomic::AtomicU64);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible

error[E0433]: failed to resolve: could not find `AtomicU64` in `atomic`
   --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:304:22
    |
304 |         Self(atomic::AtomicU64::new(self.load()))
    |                      ^^^^^^^^^
    |                      |
    |                      could not find `AtomicU64` in `atomic`
    |                      help: a struct with a similar name exists: `AtomicU16`
    |
note: struct `crate::write::zip_writer::AtomicU64` exists but is inaccessible
   --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:282:1
    |
282 | pub struct AtomicU64(atomic::AtomicU64);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible

Some errors have detailed explanations: E0412, E0433.
For more information about an error, try `rustc --explain E0412`.
```

and I don't see any place where that use can be easily avoided.  So ... This goes back to "rust or its included components really, *really* want to have access to 64-bit atomics".

bjorn3 replied:
> However, that said, the above modification concerns LLVM's reliance on 64-bit atomics, and that also needs to be dealt with one way or the other.

For that `-latomic` is fine.

> and I don't see any place where that use can be easily avoided. So ... This goes back to "rust or its included components really, really want to have access to 64-bit atomics".

Looks like those crates are hard-coding which targets lack 64-bit atomics, which used to be necessary as there was no stable way to query which atomic widths are supported, but rust 1.60.0 stabilized the `target_has_atomic` family of `#[cfg]`'s which allow querying this at compile time.

Also if I try to use `AtomicU64::compare_exchange` on i586-unknown-linux-gnu, I get an `lock cmpxchg8b`, not a call to libatomic which suggests that 64-bit atomics are supported on i586 after all.

he32 replied:
OK, it looks like you are correct wrt. `cmpxchg8b` -- it's there from Pentium onwards.  So...  I assume it's then OK to let atomic size remain at 64 for this target?  Perhaps the comment related to `-latomic` should be updated to say why it's there -- I *think* it's there because NetBSD "i386" family compilers `gcc` and `g++` are both built for i486 (under the motto "we care about backward compatibility", I'm sure), and LLVM wants/needs 64-bit atomics, ref.:

```
: {1} gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/lto-wrapper
Target: i486--netbsdelf
Configured with: /usr/src/tools/gcc/../../external/gpl3/gcc/dist/configure --target=i486--netbsdelf --enable-long-long --enable-threads --with-bugurl=http://www.NetBSD.org/Misc/send-pr.html --with-pkgversion='NetBSD nb4 20200810' --with-system-zlib --without-isl --enable-__cxa_atexit --enable-libstdcxx-time=rt --enable-libstdcxx-threads --with-diagnostics-color=auto-if-env --with-arch=i486 --with-tune=nocona --with-default-libstdcxx-abi=new --with-mpc-lib=/var/obj/mknative/i386/usr/src/external/lgpl3/mpc/lib/libmpc --with-mpfr-lib=/var/obj/mknative/i386/usr/src/external/lgpl3/mpfr/lib/libmpfr --with-gmp-lib=/var/obj/mknative/i386/usr/src/external/lgpl3/gmp/lib/libgmp --with-mpc-include=/usr/src/external/lgpl3/mpc/dist/src --with-mpfr-include=/usr/src/external/lgpl3/mpfr/dist/src --with-gmp-include=/usr/src/external/lgpl3/gmp/lib/libgmp/arch/i386 --enable-tls --disable-multilib --disable-libstdcxx-pch --build=i486--netbsdelf --host=i486--netbsdelf --with-sysroot=/var/obj/mknative/i386/usr/src/destdir.i386
Thread model: posix
gcc version 7.5.0 (nb4 20200810) 
: {2} 
: {2} uname -a
NetBSD xxxx.uninett.no 9.2 NetBSD 9.2 (GENERIC_PAE) #0: Sat Apr 16 13:06:03 CEST 2022  he@xxxx.uninett.no:/usr/obj/sys/arch/i386/compile/GENERIC_PAE i386
: {3} 
```

bjorn3 replied:
> I assume it's then OK to let atomic size remain at 64 for this target?

Looks like it. I opened an issue for crossbeam and zip to improve compatibility with other targets without 64-bit atomics, but that isn't necessary for i586-unknown-netbsd it seems.

> I think it's there because NetBSD "i386" family compilers gcc and g++ are both built for i486

That would explain it.

bjorn3 replied:
> > I think it's there because NetBSD "i386" family compilers gcc and g++ are both built for i486
> 
> That would explain it.

Could you add a comment to that effect?

he32 replied:
> > > I think it's there because NetBSD "i386" family compilers gcc and g++ are both built for i486
> >
> > That would explain it.
>
> Could you add a comment to that effect?

Done.

he32 committed:
rustc_llvm/build.rs: improve comment for NetBSD/i386 targets

...explaining why we need -latomic (gcc & g++ built for i486,
and LLVM insisting on use of 64-bit atomics).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6657498965/job/18092337672) [(plain)](https://github.com/rust-lang/rust/commit/391b472a370a5d35c43bd0a26182076cf0c17ca9/checks/18092337672/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
Removing intermediate container 99a0a9766996
 ---> e05005474f2d
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 820e715fd5fb
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-ss8irdow/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 20.4 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 47.3 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.4 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container 3eb419f8d2df
 ---> 2c373c3c1335
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> cc668459721d
 ---> cc668459721d
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 70df162fedf0
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container ec5b03b4e4aa
 ---> a90dfbeb7e84
Successfully built a90dfbeb7e84
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:a90dfbeb7e84a3bdb7934cf96979157f71e361f2c1ae21805aa7601d30e3ad58
Uploading finished image sha256:a90dfbeb7e84a3bdb7934cf96979157f71e361f2c1ae21805aa7601d30e3ad58 to https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
a90dfbeb7e84   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
cc668459721d   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
2c373c3c1335   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
820e715fd5fb   11 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
e05005474f2d   12 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      3 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      3 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      3 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f5c0f292dd0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
sha256:a90dfbeb7e84a3bdb7934cf96979157f71e361f2c1ae21805aa7601d30e3ad58
---
DirectMap4k:      257984 kB
DirectMap2M:     6033408 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/698db856de0b67313ddcb96b6599598058489ea9/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-698db856de0b67313ddcb96b6599598058489ea9-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 30.65s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_target/src/spec/i586_unknown_netbsd.rs at line 1:
-use crate::spec::{Cc, Lld, LinkerFlavor, StackProbeType, Target, TargetOptions};
+use crate::spec::{Cc, LinkerFlavor, Lld, StackProbeType, Target, TargetOptions};
 pub fn target() -> Target {
 pub fn target() -> Target {
     let mut base = super::netbsd_base::opts();
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/armv7a_none_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/hermit_base.rs" "/checkout/compiler/rustc_target/src/spec/powerpc64le_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/i586_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/aarch64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/powerpc64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/s390x_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/mips_unknown_linux_uclibc.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:43
##[error]Diff in /checkout/compiler/rustc_target/src/spec/i586_unknown_netbsd.rs at line 5:
     base.cpu = "pentium".into();
     base.max_atomic_width = Some(64);
-    base.pre_link_args
-        .entry(LinkerFlavor::Gnu(Cc::Yes, Lld::No))
-        .or_default()
-        .push("-m32".into());
+    base.pre_link_args.entry(LinkerFlavor::Gnu(Cc::Yes, Lld::No)).or_default().push("-m32".into());
     base.stack_probes = StackProbeType::Call;
     Target {
  local time: Thu Oct 26 17:18:04 UTC 2023
  network time: Thu, 26 Oct 2023 17:18:04 GMT
##[error]Process completed with exit code 1.
```

</details>

he32 commented:
> These commits modify **compiler targets**. (See the [Target Tier Policy](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html).)

Yep, this does.

>    A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)

I believe that would be me.

>    Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target.

I believe the `i586-unknown-netbsd` triple is consistent with current form.

>        Target names should not introduce undue confusion or ambiguity unless absolutely necessary to maintain ecosystem compatibility. For example, if the name of the target makes people extremely likely to form incorrect beliefs about what it targets, the name should be changed or augmented to disambiguate it.

I don't think this does.
 
>       If possible, use only letters, numbers, dashes and underscores for the name. Periods (.) are known to cause issues in Cargo.

I beleive this is compliant.

>    Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users.

This is using open source toools only.

>        The target must not introduce license incompatibilities.

No new restrictions are introduced.

>        Anything added to the Rust repository must be under the standard Rust license (MIT OR Apache-2.0).

That's agreed & understood.

>        The target must not cause the Rust tools or libraries built for any other host (even when supporting cross-compilation to the target) to depend on any new dependency less permissive than the Rust licensing policy. This applies whether the dependency is a Rust crate that would require adding new license exceptions (as specified by the tidy tool in the rust-lang/rust repository), or whether the dependency is a native library or binary. In other words, the introduction of the target must not cause a user installing or running a version of Rust or the Rust tools to be subject to any new license requirements.

There are no new dependencies/features required.

>        Compiling, linking, and emitting functional binaries, libraries, or other code for the target (whether hosted on the target itself or cross-compiling from another target) must not depend on proprietary (non-FOSS) libraries. Host tools built for the target itself may depend on the ordinary runtime libraries supplied by the platform and commonly used by other applications built for the target, but those libraries must not be required for code generation for the target; cross-compilation to the target must not require such libraries at all. For instance, rustc built for the target may depend on a common proprietary C runtime library or console output library, but must not depend on a proprietary code generation library or code optimization library. Rust's license permits such combinations, but the Rust project has no interest in maintaining such combinations within the scope of Rust itself, even at tier 3.

As previously said it's using open source tools only.

>        "onerous" here is an intentionally subjective term. At a minimum, "onerous" legal/licensing terms include but are not limited to: non-disclosure requirements, non-compete requirements, contributor license agreements (CLAs) or equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional on the employer or employment of any particular Rust developers, revocable terms, any requirements that create liability for the Rust project or its developers or users, or any requirements that adversely affect the livelihood or prospects of the Rust project or its developers or users.

There are no such terms present.

>    Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions.

I'm not the reviewer here.

>        This requirement does not prevent part or all of this policy from being cited in an explicit contract or work agreement (e.g. to implement or maintain support for a target). This requirement exists to ensure that a developer or team responsible for reviewing and approving a target does not face any legal threats or obligations that would prevent them from freely exercising their judgment in such approval, even if such judgment involves subjective matters or goes beyond the letter of these requirements.

Again I'm not the reviewer here.

>    Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate (core for most targets, alloc for targets that can support dynamic memory allocation, std for targets with an operating system or equivalent layer of system-provided functionality), but may leave some code unimplemented (either unavailable or stubbed out as appropriate), whether because the target makes it impossible to implement or challenging to implement. The authors of pull requests are not obligated to avoid calling any portions of the standard library on the basis of a tier 3 target not implementing those portions.
>    The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible. If the target supports running binaries, or running tests (even if they do not pass), the documentation must explain how to run such binaries or tests for the target, using emulation if possible or dedicated hardware if necessary.

Building is described in platform support doc, there is (perhaps in error...) already an entry for this target there.  I admit to not having tried running the tests; and instead abuse rust itself as its own self-test ("it builds, ship it!").  Well, not entirely true, I verify among other things that `firefox` builds and runs.  This action (adding this target) should forestall its scheduled removal from the target description, ref. the recent action to clean up those.

>    Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via @) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages.

Understood.

>        Backlinks such as those generated by the issue/PR tracker when linking to an issue or PR are not considered a violation of this policy, within reason. However, such messages (even on a separate repository) must not generate notifications to anyone involved with a PR who has not requested such notifications.

Understood.

>    Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target.

I believe I didn't break any other target.

>        In particular, this may come up when working on closely related targets, such as variations of the same architecture with different features. Avoid introducing unconditional uses of features that another variation of the target may not have; use conditional compilation or runtime detection, as appropriate, to let each target run code supported by that target.

I think there are no such problems in this PR.

bjorn3 commented:
Thanks!

@bors r+

TaKO8Ki commented:
It seems that this pull request causes an error on CI. @bors r -

https://github.com/rust-lang/rust/pull/117255#issuecomment-1782176406

TaKO8Ki commented:
It seems that this pull request causes an error on CI. @bors r-

https://github.com/rust-lang/rust/pull/117255#issuecomment-1782176406

he32 committed:
i586_unknown_netbsd.rs: fix formatting.

This hopefully fixes the CI run after integration of this
target.

he32 commented:
> It seems that this pull request causes an error on CI. @bors r-
> 
> [#117255 (comment)](https://github.com/rust-lang/rust/pull/117255#issuecomment-1782176406)

This appears to concern the formatting.  I've pushed a commit which should fix the two flagged issues (sort of "use" content, and formatting).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6664193547/job/18111352944) [(plain)](https://github.com/rust-lang/rust/commit/893e7266376f8c7629a5b5130e95bc3063128f77/checks/18111352944/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
Removing intermediate container 1998b6f9d066
 ---> c9d9a530a174
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 9052d773b77c
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-88yk2p4b/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 60.0 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.4 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container 4ef31f008d5e
 ---> 2256749b1f0d
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 080005c28d2c
 ---> 080005c28d2c
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> e040951830b6
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 363b71113302
 ---> 36711023e5c7
Successfully built 36711023e5c7
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:36711023e5c790379bb2355a566be8facd14c95d6c0943fb612c4fba1670745f
Uploading finished image sha256:36711023e5c790379bb2355a566be8facd14c95d6c0943fb612c4fba1670745f to https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
IMAGE          CREATED              CREATED BY                                      SIZE      COMMENT
36711023e5c7   1 second ago         /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
080005c28d2c   2 seconds ago        /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
2256749b1f0d   3 seconds ago        |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
9052d773b77c   10 seconds ago       /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
c9d9a530a174   11 seconds ago       |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      3 weeks ago          /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      3 weeks ago          /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      3 weeks ago          /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7ffa1adb6e50>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
sha256:36711023e5c790379bb2355a566be8facd14c95d6c0943fb612c4fba1670745f
---
DirectMap4k:      272320 kB
DirectMap2M:     7067648 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/707d8c3f1bba10d5aaa9c8719c410711edf2e227/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-707d8c3f1bba10d5aaa9c8719c410711edf2e227-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
extracting /checkout/obj/build/cache/llvm-707d8c3f1bba10d5aaa9c8719c410711edf2e227-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
##[group]Building stage0 tool tidy (x86_64-unknown-linux-gnu)
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_llvm/build.rs at line 261:
         // On NetBSD/i386, gcc and g++ is built for i486 (to maximize backward compat)
         // However, LLVM insists on using 64-bit atomics.
         // This gives rise to a need to link rust itself with -latomic for these targets
-        if target.starts_with("i586")
-           || target.starts_with("i686")
-        {
+        if target.starts_with("i586") || target.starts_with("i686") {
             println!("cargo:rustc-link-lib=atomic");
         }
         println!("cargo:rustc-link-lib=z");
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_index/src/bit_set/tests.rs" "/checkout/compiler/rustc_llvm/build.rs" "/checkout/compiler/rustc_index/src/interval/tests.rs" "/checkout/compiler/rustc_llvm/src/lib.rs" "/checkout/compiler/rustc_fluent_macro/src/fluent.rs" "/checkout/compiler/rustc_fluent_macro/src/lib.rs" "/checkout/compiler/rustc_hir_analysis/src/bounds.rs" "/checkout/compiler/rustc_index/src/vec/tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Oct 27 07:30:33 UTC 2023
  network time: Fri, 27 Oct 2023 07:30:33 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

he32 committed:
For i586/NetBSD: fix another formatting insistence.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6665517274/job/18115246311) [(plain)](https://github.com/rust-lang/rust/commit/0f04e2dd8f08951b3b9d0641febdcf3003f876aa/checks/18115246311/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- spec::tests::i586_unknown_netbsd stdout ----
thread 'spec::tests::i586_unknown_netbsd' panicked at compiler/rustc_target/src/spec/tests/tests_impl.rs:85:17:
error: test failed, to rerun pass `-p rustc_target --lib`
assertion `left == right` failed
  left: Some(["-m32"])
 right: None


failures:
    spec::tests::i586_unknown_netbsd
```

</details>

he32 committed:
i586_unknown_netbsd.rs: drop "-m32" flag insertion to gcc.

This triggers a consistency check in rust (that all linker flavours
must have identical arguments), and on NetBSD/i386, the 32-bitness
is implicitly chosen through the chosen toolchain, and appears to
not be required.  So drop it, and also drop the imports of the
now-no-longer-used identifiers.

bjorn3 commented:
> error: target `i586-unknown-netbsd` is missing from platform-support.md

I think you need to add it to the platform-support.md index.

he32 committed:
i586-unknown-netbsd: add entry in platform-support.md.

he32 commented:
>> error: target `i586-unknown-netbsd` is missing from platform-support.md
>
> I think you need to add it to the platform-support.md index.

Yep, figured that out, and just did so; awaiting new build/check
results.

bjorn3 commented on src/doc/rustc/src/platform-support.md:
```diff
@@ -152,6 +152,7 @@ target | std | notes
 `i586-pc-windows-msvc` | * | 32-bit Windows w/o SSE [^x86_32-floats-x87]
 `i586-unknown-linux-gnu` | ✓ | 32-bit Linux w/o SSE (kernel 3.2, glibc 2.17) [^x86_32-floats-x87]
 `i586-unknown-linux-musl` | ✓ | 32-bit Linux w/o SSE, MUSL [^x86_32-floats-x87]
+[`i586-unknown-netbsd`](platform-support/netbsd.md) |  ✓ | 32-bir x86, restricted to Pentium
```
*bit

he32 replied:
How embarassing...  Fixed.

he32 committed:
i586-unknown-netbsd platform-support.md: fix typo.

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit f5fa36fbb72fe6d676adcb9bb7c1cba869c81c00 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 f5fa36fbb72fe6d676adcb9bb7c1cba869c81c00 -->
<!-- homu: {"type":"Approved","sha":"f5fa36fbb72fe6d676adcb9bb7c1cba869c81c00","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented on src/doc/rustc/src/platform-support.md:
```diff
@@ -152,6 +152,7 @@ target | std | notes
 `i586-pc-windows-msvc` | * | 32-bit Windows w/o SSE [^x86_32-floats-x87]
 `i586-unknown-linux-gnu` | ✓ | 32-bit Linux w/o SSE (kernel 3.2, glibc 2.17) [^x86_32-floats-x87]
 `i586-unknown-linux-musl` | ✓ | 32-bit Linux w/o SSE, MUSL [^x86_32-floats-x87]
+[`i586-unknown-netbsd`](platform-support/netbsd.md) |  ✓ | 32-bit x86, restricted to Pentium
```
Could you please add a footnote for the usual x86-32 float problem? See the other i686/i586 targets. Depending on whether this target has SSE or not, you need `x86_32-floats-return-ABI` or `x86_32-floats-x87`.
